internal/graph/check.go

```diff
@@ -282,6 +283,9 @@ func union(ctx context.Context, concurrencyLimit uint32, handlers ...CheckHandle
 		}
 	}
 
+	if err != nil {
+		telemetry.TraceError(trace.SpanFromContext(ctx), err)
```

Comment:
IMO these shouldn't leak outside of the function defining the span itself. If a functionA creates a span and has multiple functions it calls functionB, functionC, etc.. and if any of called functions return an error, then the caller functionA should record the error event.

I believe we should prioritize more `defer` based capture here to avoid code smells having this scattered everywhere.

e.g.

```go
func functionA(ctx context.Context) ... {
    ctx, span := tracer.Start(ctx, "functionA")
    
    var err error

    defer func() {
        span.SetError(err) // only sets the error event if err is non-nil
        span.End()
    }()
   
    err = functionB(ctx)
    err = functionC(ctx)
}
```

Reply:
I tried this change inside `checkSetOperation`:

```
// ...
return func(ctx context.Context) (*ResolveCheckResponse, error) {
		var err error
		var resp *ResolveCheckResponse

		ctx, span := tracer.Start(ctx, reducerKey)
		defer func() {
			span.End()
			if err != nil {
				telemetry.TraceError(span, err)
			}
		}()

		resp, err = reducer(ctx, c.concurrencyLimit, handlers...)
		return resp, err
	}
```

But it doesn't propagate all the way to the root span:

<img width="1428" alt="image" src="https://github.com/openfga/openfga/assets/5374887/851b8992-1f65-4cfb-81a8-adf5d32e944b">

I'll spend some time later to see if it can be improved

Reply:
I'm an idiot. I put the `span.RecordError(...)` _after_ the `span.End()`